cluster-autoscaler/clusterstate/clusterstate.go

```diff
@@ -154,21 +154,39 @@ func NewClusterStateRegistry(cloudProvider cloudprovider.CloudProvider, config C
 	}
 }
 
-// RegisterScaleUp registers scale up.
-func (csr *ClusterStateRegistry) RegisterScaleUp(request *ScaleUpRequest) {
+// RegisterOrUpdateScaleUp registers scaleup for give node group or changes requested node increase
+// count.
+// If increaseDelta is positive then number of new nodes requested is increased; Time and expectedAddTime
+// are reset.
+// If increaseDelete is negative the number of new nodes requested is decreased; Time and expectedAddTime are
+// left intact.
+func (csr *ClusterStateRegistry) RegisterOrUpdateScaleUp(nodeGroupName string, increaseDelta int) {
 	csr.Lock()
 	defer csr.Unlock()
 
-	oldScaleUpRequest, found := csr.scaleUpRequests[request.NodeGroupName]
+	scaleUpRequest, found := csr.scaleUpRequests[nodeGroupName]
 	if !found {
-		csr.scaleUpRequests[request.NodeGroupName] = request
-		return
+		scaleUpRequest = &ScaleUpRequest{
+			NodeGroupName:   nodeGroupName,
+			Time:            time.Now(),
```

Comment:
Hard-coding time.Now() here seems like a step backwards - we've been reducing it across the codebase, both to make unit testing easier and the reported behavior more consistent.

Note that the previous change was implementation-only and didn't require updating any unit tests - they continued to work and helped us verify that the refactoring indeed didn't change the behavior in covered scenarios. In other words, they weren't "change-detection" tests. With this change, you had to add manipulating internal state of ClusterStateRegistry (substituting timestamps) for them to continue to work. This means that:
1. if we were to revert the other change, or made another refactoring of internal implementation, it would break the tests this time, because they now make assumptions about internal state and manipulate it,
2. old tests no longer verify the logic of aggregating scale-up requests and using a correct timeout - we replace these values. If you set ExpectedAddTime to time.Now() in this method, backoff tests wouldn't fail. So a new test had to be added to cover this, making even more assumptions about internal state (which makes subsequent code changes harder, see (1)). 

(Yes, a lot of our unit tests make assertions on internal state, I'm aware of this;P)

Reply:
> Hard-coding time.Now() here seems like a step backwards - we've been reducing it across the codebase, both to make unit testing easier and the reported behavior more consistent.

The problem here is using static `time.Now()` method. We could fix that if we passed some `Watch` object on construction of `ClusterStateRegistry`. Standard implementation would represent wall time. And in tests we could put sth where we can manipulate the time programtically. Not sure if golang have sth like this.

> With this change, you had to add manipulating internal state of ClusterStateRegistry

I agree that I do that. Though I disagree that these tests did not do that already. Just the internal state was exposed to the outer caller via `ScaleUpRequest` struct.

Reply:
> I faintly recall seeing some flags and configuration options (i.e. things normally not exposed to GKE users) getting moved into ConfigMaps. I don't know the details, but perhaps this approach would some day work for tuning the autoscaler in GKE (although I'm sure that further complicates the GKE support story).

Technically, we could use an internal timeNow (or sth like this) function, assign time.Now() to it in prod and custom fakeTime() in tests. This complicates test setup, though.

> Just the internal state was exposed to the outer caller via ScaleUpRequest struct.

To me, it seems we passed input parameter and then never checked how CSR stored it, only that it made correct assessments based on it. 

Would it be OK to pass startTime and expectedAddDuration (or expectedAddTime, doesn't matter) as parameters directly? Or if ScaleUpRequest (exported) was part of this method's interface, but internally another type was used for storing scale-up state?